### DISCUSSION TITLE
setting local part coordinate system for UR robot

### DISCUSSION THREAD

**Discussion by jco-c** [visose/Robots#182] [Q&A] [answered]
Hi,

for UR robots (and for others as well i think) there is the possibility to set a local coordinate system for the part which is worked on (UR calls this local PCS). This allows to teach the position and orientation of the part to the robot, to allow for varying part positions. Is there a possibility to add this to the program which is created in robots?

**Comment by visose** [comment]
Yes, there is a Frame component for this. When you create a target, right click and select Frame to get an input for it.

Currently in the generated code for UR this transformation is baked in, but I'll change it in the future so that it shows as a variable.

**Reply by jco-c** [reply]
Oh that's great! I just tried it and it works correctly in the simulation in grasshopper, but I can't find a way to set the local PCS in my UR teach pendant (UR5e) for something else than GCode. Do you have a hint for me, how to make the frames work with UR robots?

**Reply by visose** [reply]
I guess you are using the `The Remote TCP & Toolpath URCap Controller` to set the local coordinate system. The Robots plugin doesn't support functionality from the module processpath. I'm not sure if this will be added, as it has to support older CB2 robots.

For the time being, I will change the generated code so that the frame (local coord system) is a variable that can be changed, and then use either `pose_trans` or `path_offset_set`.

If the frame data is set to be read from the controller, rather than from Grasshopper (this is done by defining a Frame element in the xml of your robot library), it will use an installation variable that gets stored in the controller separate from the program.

**Reply by jco-c** [reply]
Yes, I was trying to use the Toolpath function. 

Thank you for your effort! So starting with the next version I would be able to modify the variable in the installation tab of the program? And do I need to modify the xml myself? Really looking forward to this! If there is anything I need to change in the gh program to make this work, a short instruction would be greatly appreciated :)

**Reply by visose** [reply]
Not sure if the next update or the one after.
1. If you add a frame using the Create Frame component, the variable will apear in the generated code.
1. If you add a frame using the Load Frame component, for it to show up in the list, you'll need to add it to the xml of your robot library like this:
```
<Frame name="FrameName">
    <Base x="0" y="0" z="0" q1="1" q2="0" q3="0" q4="0" />
</Frame>
```
It will then try to use a variable from the controller that has the same name. The values in the xml will just be used for the simulation in Grasshopper. They don't have to match the ones in the controller exactly, but the closer they are, the closer the simulation will be.

**Comment by visose** [comment, marked as answer]
Closing this as answered. For Robots, local part coordinate systems are handled with Frames: add the Frame input to Create Target, or define a Frame in the robot XML and load it. UR processpath/Remote TCP & Toolpath URCap-specific functionality is not directly supported, so any controller-side frame variable needs to be represented through the generated frame variable or matching XML/controller setup.
